File: haskell/glue/src/Glue/Lib/Math/Utility/Abs.hs
module Glue.Lib.Math.Utility.Abs where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Absolute value function
-- Mirrors Haskell Glue.Lib.Math.Utility.Abs.abs exactly
abs :: IR Eval
abs = NativeFunc absImpl

-- Absolute value function implementation
-- Mirrors Haskell Glue.Lib.Math.Utility.Abs.absImpl exactly
absImpl :: IR Eval -> Eval (IR Eval)
absImpl arg = case arg of
    Integer n -> pure $ Integer (Prelude.abs n)
    Float n -> pure $ Float (Prelude.abs n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/utility/abs.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Absolute value function
/// Mirrors Haskell Glue.Lib.Math.Utility.Abs.abs exactly
final Ir abs = IrNativeFunc(absImpl);

/// Absolute value function implementation
/// Mirrors Haskell Glue.Lib.Math.Utility.Abs.absImpl exactly
Eval<Ir> absImpl(Ir arg) {
  return switch (arg) {
    IrInteger(value: final n) => Eval.pure(IrInteger(n.abs())),
    IrFloat(value: final n) => Eval.pure(IrFloat(n.abs())),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
